crates/typst/src/export/pdf/gradient.rs

```diff
@@ -0,0 +1,197 @@
+};
+
+use super::color::{CSFunctions, PaintEncode};
+use super::page::{PageContext, Transforms};
+use super::{PdfContext, RefExt};
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub(super) struct PdfGradient {
+    /// The transform to apply to the gradient.
+    pub transform: Transform,
+
+    /// The size of the gradient.
+    /// Required for aspect ratio correction.
+    pub size: Size,
+
+    /// The gradient.
+    pub gradient: Gradient,
+}
+
+pub fn write_gradients(ctx: &mut PdfContext) {
+    for PdfGradient { transform, size, gradient } in
+        ctx.gradient_map.items().cloned().collect::<Vec<_>>()
+    {
+        let shading = ctx.alloc.bump();
+        ctx.gradient_refs.push(shading);
+
+        let mut shading_pattern = match &gradient {
+            Gradient::Linear(linear) => {
+                let shading_function = shading_function(ctx, &gradient);
+                let mut shading_pattern = ctx.writer.shading_pattern(shading);
+                let mut shading = shading_pattern.function_shading();
+                shading.shading_type(FunctionShadingType::Axial);
+
+                ctx.colors
+                    .write(gradient.space(), shading.color_space(), &mut ctx.alloc);
+
+                let angle = linear.angle.correct_aspect_ratio(size);
+                let (sin, cos) = angle.sin_cos();
+                let length = sin.abs() + cos.abs();
+
+                shading
+                    .anti_alias(gradient.anti_alias())
+                    .function(shading_function)
+                    .coords([0.0, 0.0, length as f32, 0.0])
+                    .extend([true; 2]);
+
+                shading.finish();
+
+                shading_pattern
+            }
+        };
+
+        shading_pattern.matrix(transform.as_array());
+        shading_pattern.finish();
+    }
+}
+
+fn shading_function(ctx: &mut PdfContext, gradient: &Gradient) -> Ref {
+    let function = ctx.alloc.bump();
+    let mut functions = vec![];
+    let mut bounds = vec![];
+    let mut encode = vec![];
+
+    // Create the individual gradient functions for each pair of stops
+    for window in gradient.stops().windows(2) {
+        let (first, second) = (window[0], window[1]);
+
+        // Skip stops with the same position
+        if first.offset.unwrap().get() == second.offset.unwrap().get() {
+            continue;
+        }
+
+        bounds.push(second.offset.unwrap().get() as f32);
+        functions.push(single_gradient(
+            ctx,
+            None,
+            first.color,
+            second.color,
+            gradient.space(),
+        ));
+        encode.extend([0.0, 1.0]);
+    }
+
+    // Special case for gradients with only two stops
+    if functions.len() == 1 {
+        return functions[0];
+    }
+
+    // Remove the last bound, since it's not needed for the stitching function
+    bounds.pop();
+
+    // Create the stitching function
```

Comment:
```suggestion
    // Special case for gradients with only two stops.
    if functions.len() == 1 {
        return functions[0];
    }

    // Remove the last bound, since it's not needed for the stitching function.
    bounds.pop();

    // Create the stitching function.
```